Feature hcc profile

bensander opened the pull request:
Simple but useful GPU command profiler for HCC.  Enabled with env var, shows kernel and data  xfer times and bandwidth.

bensander committed:
Add simple profile mode for kernels and memory copies.

- HCC_PROFILE=1 env var
- prints one-line message with kernels and copy commands + execution
  time on GPU.
- dispatch_hsa_kernel now accepts an optional kernel name

- in syncCopyExt, don't release the signal - this should occur when copy
  is disposed and not before

Other debug and profiling cleanup and improvements:

- Add DB_CODE to help resolve CreateKernel mysteries, set HCC_DB=0x4000
- Create a short function name for kernels.  Not guaranteed to be unique
  but easier for humans to parse.  Used in kernel profiling.
- Improved debug messages for CreateKernel
- Refactor getKernelName().
- Correctly collect async copy timestamps in get*TimeStamp.
- Track kind of the copy (isAsync, isSingleStepCopy, isPeerToPeer)
  for future debug display.
- Separate full AQL dumps into DB_AQL2.  DB_AQL now just shows
  the decoded packet.
- add hccgetenv for reading string env vars.
- Add 0x for some signal values that were printed in hex.

bensander committed:
Update cmdline profiler.

- For multi-step copies (ie pin-in-place or staging), store API
  start-time.
- tweak kernel short-name calculaton.
- add switch to support tracing barriers.
    - Note time shown for barrier includes time to wait for dependencies
      to resolve.

aaronenyeshi commented:
Hi, is there documentation on these debuggers and profilers? I think it would be very useful to add how to use this tool (such as which env var to set) into debugging documentations.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -41,6 +41,7 @@
 #include <time.h>
 #include <iomanip>
 
+#define KALMAR_DEBUG (0)
```
@bensander please help remove this line.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -2203,9 +2276,68 @@ class HSADevice final : public KalmarDevice
     }
 
     void* CreateKernel(const char* fun, Kalmar::KalmarQueue *queue) override {
+        //fun = "_ZZN8hip_impl21grid_launch_hip_impl_IZ76Cijk_Alik_Bljk_SB_MT064x128x08_GSU06_MO10_PGR0_PLR0_TT04_08_WG16_16_01_WGM01PfPKfS3_ffjjjjjjjjjjjjjP12ihipStream_tjPP11ihipEvent_tS8_E4$_36JNS_17Empty_launch_parmERS1_RjSC_SC_SC_SC_SC_EEEvNS_12_GLOBAL__N_119New_grid_launch_tagE4dim3SF_iRKN2hc16accelerator_viewET_DpOT0_ENUlRKNSG_11tiled_indexILi3EEEE_19__cxxamp_trampolineES1_jjjjjj";
```
@bensander please help remove this line.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -105,6 +106,31 @@ unsigned HCC_DB = 0;
+
+int HCC_PROFILE=0;
+int HCC_PROFILE_VERBOSE=0;
+char * HCC_PROFILE_FILE=nullptr;
```
@bensander i'm seeing more file-scope global variables growing... how about we move them into `HSAContext`?

bensander replied:
Yes - could be separate pull request?  these are environment variables which arguably are best as globals.  the file handle should be moved.

whchung replied:
ok to make them in another PR.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -395,6 +429,10 @@ class HSACopy : public Kalmar::KalmarAsyncOp {
     hsa_signal_t signal;
     int signalIndex;
     bool isSubmitted;
+    bool isAsync;;          // copy was performed asynchronously
```
@bensander double `;` in the code.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -4509,8 +4751,6 @@ HSACopy::syncCopyExt(hc::hcCommandKind copyDir, const hc::AmPointerInfo &srcPtrI
             DBOUT(DB_COPY, "HSACopy::syncCopy(), hsa_amd_memory_async_copy() returns: 0x" << std::hex << hsa_status << std::dec <<"\n");
             throw Kalmar::runtime_exception("hsa_amd_memory_async_copy error", hsa_status);
         }
-        Kalmar::ctx.releaseSignal(signal, signalIndex);
```
@bensander could you provide justification why these 2 lines be removed in this PR? I thought this PR is to add kernel profiling logic. but changing these 2 lines would essentially change the behavior of HSA signal reclaim process. I'm not alluding the code is wrong, but perhaps it should be made in a separate PR?

bensander replied:
I believe this is not the right place to deallocate a single - instead this should occur in the HSACopy::dispose() routine.  The new profile mode is implemented inside the dispose routine and uses the timestamps from the signal so requires this change.

whchung replied:
thanks for the clarification

bensander committed:
Improve pull request based on review feedback:

Move g_hccProfile* variables into context.
Move ReadEnv into HSAContext variable.
Remove some debug or temp lines that should be removed.
